File: WebFt30/src/app/core/store/mapping/ft-enums-mapping.ts
import * as grpc from 'src/grpc-generated';
import {
  ChannelEvent,
  EventStatus,
  FiberState,
  MonitoringCurrentStep,
  MonitoringState,
  OpticalAccidentType,
  RtuMaker,
  RtuPartState,
  TceLinkState
} from '../models/ft30/ft-enums';
import { ReturnCode } from '../models/ft30/return-code';

export class FtEnumsMapping {
  static fromGrpcRtuMaker(grpcRtuMaker: grpc.RtuMaker): RtuMaker {
    switch (grpcRtuMaker) {
      case grpc.RtuMaker.Iit:
        return RtuMaker.Iit;
      case grpc.RtuMaker.Veex:
        return RtuMaker.Veex;
      default:
        throw new Error(`Unknown grpc.RtuMaker: ${grpcRtuMaker}`);
    }
  }

  static fromGrpcChannelEvent(grpcChannelEvent: grpc.ChannelEvent): ChannelEvent {
    // prettier-ignore
    switch(grpcChannelEvent) {
      case grpc.ChannelEvent.ChannelEvent_Broken: return ChannelEvent.Broken;
      case grpc.ChannelEvent.ChannelEvent_Nothing: return ChannelEvent.Nothing;
      case grpc.ChannelEvent.Repaired: return ChannelEvent.Repaired;
      default:
        throw new Error(`Unknown grpc.ChannelEvent: ${grpcChannelEvent}`);  }
  }

  static fromGrpcFiberState(grpcFiberState: grpc.FiberState): FiberState {
    // prettier-ignore
    switch (grpcFiberState) {
      case grpc.FiberState.NotInTrace: return FiberState.NotInTrace;
      case grpc.FiberState.NotJoined: return FiberState.NotJoined;
      case grpc.FiberState.FiberState_Unknown: return FiberState.Unknown;
      case grpc.FiberState.NotInZone: return FiberState.NotInZone;
      case grpc.FiberState.FiberState_Ok: return FiberState.Ok;
      case grpc.FiberState.Suspicion: return FiberState.Suspicion;
      case grpc.FiberState.Minor: return FiberState.Minor;
      case grpc.FiberState.Major: return FiberState.Major;
      case grpc.FiberState.Critical: return FiberState.Critical;
      case grpc.FiberState.User: return FiberState.User;
      case grpc.FiberState.FiberBreak: return FiberState.FiberBreak;
      case grpc.FiberState.NoFiber: return FiberState.NoFiber;
      case grpc.FiberState.HighLighted: return FiberState.HighLighted;
      case grpc.FiberState.DistanceMeasurement: return FiberState.DistanceMeasurement;
      case grpc.FiberState.Nothing: return FiberState.Nothing;
      default:
        throw new Error(`Unknown grpc.FiberState: ${grpcFiberState}`);
    }
  }

  static toGrpcFiberState(state: FiberState): grpc.FiberState {
    const enumCode = this.getFiberStateNumericKey(state);
    return enumCode;
  }

  static getFiberStateNumericKey(enumValue: FiberState): number {
    return +Object.keys(FiberState)
      .filter((k) => !isNaN(Number(k)))
      .find((g) => FiberState[+g] === FiberState[enumValue])!;
  }

  static getReturnCodeNumericKey(enumValue: ReturnCode): number {
    return +Object.keys(ReturnCode)
      .filter((k) => !isNaN(Number(k)))
      .find((g) => ReturnCode[+g] === ReturnCode[enumValue])!;
  }

  static getGrpcReturnCodeNumericKey(enumValue: grpc.ReturnCode): number {
    return +Object.keys(grpc.ReturnCode)
      .filter((k) => !isNaN(Number(k)))
      .find((g) => grpc.ReturnCode[+g] === grpc.ReturnCode[enumValue])!;
  }

  static fromGrpcReturnCode(grpcReturnCode: grpc.ReturnCode): ReturnCode {
    const enumCode = this.getReturnCodeNumericKey(grpcReturnCode);
    return enumCode;
  }

  static fromGrpcMonitoringState(grpcMonitoringState: grpc.MonitoringState): MonitoringState {
    // prettier-ignore
    switch (grpcMonitoringState) {
        case grpc.MonitoringState.MonitoringState_Unknown: return MonitoringState.Unknown;
        case grpc.MonitoringState.Off: return MonitoringState.Off;
        case grpc.MonitoringState.On: return MonitoringState.On;
        default:
            throw new Error(`Unknown grpc.MonitoringState: ${grpcMonitoringState}`);
    }
  }

  static fromGrpcRtuPartState(grpcRtuPartState: grpc.RtuPartState): RtuPartState {
    // prettier-ignore
    switch (grpcRtuPartState) {
        case grpc.RtuPartState.Broken: return RtuPartState.Broken;
        case grpc.RtuPartState.NotSetYet: return RtuPartState.NotSetYet;
        case grpc.RtuPartState.RtuPartState_Ok: return RtuPartState.Ok;
        default:
            throw new Error(`Unknown grpc.RtuPartState: ${grpcRtuPartState}`);
    }
  }

  static fromGrpcTceLinkState(grpcTceLinkState: grpc.TceLinkState): TceLinkState {
    // prettier-ignore
    switch(grpcTceLinkState) {
      case grpc.TceLinkState.NoLink: return TceLinkState.NoLink;
      case grpc.TceLinkState.SnmpTrapOff: return TceLinkState.SnmpTrapOff;
      case grpc.TceLinkState.SnmpTrapOn: return TceLinkState.SnmpTrapOn;
      default:
        throw new Error(`Unknown grpc.TceLinkState: ${grpcTceLinkState}`);
    }
  }

  // можно присваивать напрямую
  // static fromGrpcBaseRefType(grpcType: grpc.BaseRefType): BaseRefType {
  //   // prettier-ignore
  //   switch(grpcType) {
  //     case grpc.BaseRefType.None: return BaseRefType.None;
  //     case grpc.BaseRefType.Precise: return BaseRefType.Precise;
  //     case grpc.BaseRefType.Fast: return BaseRefType.Fast;
  //     case grpc.BaseRefType.Additional: return BaseRefType.Additional;
  //     default:
  //       throw new Error(`Unknown grpc.BaseRefType: ${grpcType}`);
  //   }
  // }

  static fromGrpcEventStatus(grpcStatus: grpc.EventStatus): EventStatus {
    // prettier-ignore
    switch(grpcStatus) {
      case grpc.EventStatus.JustMeasurementNotAnEvent: return EventStatus.JustMeasurementNotAnEvent;
      case grpc.EventStatus.EventButNotAnAccident: return EventStatus.EventButNotAnAccident;
      case grpc.EventStatus.Unprocessed: return EventStatus.Unprocessed;
      case grpc.EventStatus.NotImportant: return EventStatus.NotImportant;
      case grpc.EventStatus.Planned: return EventStatus.Planned;
      case grpc.EventStatus.NotConfirmed: return EventStatus.NotConfirmed;
      case grpc.EventStatus.Suspended: return EventStatus.Suspended;
      case grpc.EventStatus.Confirmed: return EventStatus.Confirmed;
      default:
        throw new Error(`Unknown grpc.EventStatus: ${grpcStatus}`);
    }
  }

  static fromGrpcCurrentStep(grpcStep: grpc.MonitoringCurrentStep): MonitoringCurrentStep {
    // prettier-ignore
    switch(grpcStep) {
      case grpc.MonitoringCurrentStep.Unknown: return MonitoringCurrentStep.Unknown;
      case grpc.MonitoringCurrentStep.Idle: return MonitoringCurrentStep.Idle;
      case grpc.MonitoringCurrentStep.Toggle: return MonitoringCurrentStep.Toggle;
      case grpc.MonitoringCurrentStep.Measure: return MonitoringCurrentStep.Measure;
      case grpc.MonitoringCurrentStep.FailedOtauProblem: return MonitoringCurrentStep.FailedOtauProblem;
      case grpc.MonitoringCurrentStep.FailedOtdrProblem: return MonitoringCurrentStep.FailedOtdrProblem;
      case grpc.MonitoringCurrentStep.Interrupted: return MonitoringCurrentStep.Interrupted;
      case grpc.MonitoringCurrentStep.Analysis: return MonitoringCurrentStep.Analysis;
      case grpc.MonitoringCurrentStep.MeasurementFinished: return MonitoringCurrentStep.MeasurementFinished;
      default:
        throw new Error(`Unknown grpc.MonitoringCurrentStep: ${grpcStep}`);
    }
  }

  static fromGrpcOpticalAccidentType(grpcType: grpc.OpticalAccidentType): OpticalAccidentType {
    // prettier-ignore
    switch(grpcType) {
      case grpc.OpticalAccidentType.Break: return OpticalAccidentType.Break;
      case grpc.OpticalAccidentType.Loss: return OpticalAccidentType.Loss;
      case grpc.OpticalAccidentType.Reflectance: return OpticalAccidentType.Reflectace;
      case grpc.OpticalAccidentType.LossCoeff: return OpticalAccidentType.LossCoef;
      case grpc.OpticalAccidentType.TotalLoss: return OpticalAccidentType.TotalLoss;
      case grpc.OpticalAccidentType.OpticalAccidentType_None: return OpticalAccidentType.None;
    }
  }
}

File: Backend/src/Fibertest30.Api/Mapping/FtEnumsMapping.cs
using Iit.Fibertest.Dto;

namespace Fibertest30.Api;

public static class FtEnumsMapping
{
    public static EquipmentType ToProto(this Iit.Fibertest.Dto.EquipmentType type)
    {
        return type switch
        {
            Iit.Fibertest.Dto.EquipmentType.AdjustmentPoint => EquipmentType.AdjustmentPoint,
            Iit.Fibertest.Dto.EquipmentType.EmptyNode => EquipmentType.EmptyNode,
            Iit.Fibertest.Dto.EquipmentType.CableReserve => EquipmentType.CableReserve,
            Iit.Fibertest.Dto.EquipmentType.Other => EquipmentType.Other,
            Iit.Fibertest.Dto.EquipmentType.Closure => EquipmentType.Closure,
            Iit.Fibertest.Dto.EquipmentType.Cross => EquipmentType.Cross,
            Iit.Fibertest.Dto.EquipmentType.Well => EquipmentType.Well,
            Iit.Fibertest.Dto.EquipmentType.Terminal => EquipmentType.Terminal,
            Iit.Fibertest.Dto.EquipmentType.Rtu => EquipmentType.Rtu,
            Iit.Fibertest.Dto.EquipmentType.AccidentPlace => EquipmentType.AccidentPlace,
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    public static Iit.Fibertest.Dto.EquipmentType FromProto(this EquipmentType type)
    {
        var num = (int)type;
        return (Iit.Fibertest.Dto.EquipmentType)num;
    }

    public static RtuMaker ToProto(this Iit.Fibertest.Dto.RtuMaker rtuMaker)
    {
        return rtuMaker switch
        {
            Iit.Fibertest.Dto.RtuMaker.IIT => RtuMaker.Iit,
            Iit.Fibertest.Dto.RtuMaker.VeEX => RtuMaker.Veex,
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    public static Iit.Fibertest.Dto.RtuMaker FromProto(this RtuMaker rtuMaker)
    {
        int code = (int)rtuMaker;
        return (Iit.Fibertest.Dto.RtuMaker)code;
    }


    public static RtuPartState ToProto(this Iit.Fibertest.Dto.RtuPartState state)
    {
        return state switch
        {
            Iit.Fibertest.Dto.RtuPartState.Broken => RtuPartState.Broken,
            Iit.Fibertest.Dto.RtuPartState.NotSetYet => RtuPartState.NotSetYet,
            Iit.Fibertest.Dto.RtuPartState.Ok => RtuPartState.Ok,
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    public static MonitoringState ToProto(this Iit.Fibertest.Dto.MonitoringState state)
    {
        return state switch
        {
            Iit.Fibertest.Dto.MonitoringState.Unknown => MonitoringState.Unknown,
            Iit.Fibertest.Dto.MonitoringState.Off => MonitoringState.Off,
            Iit.Fibertest.Dto.MonitoringState.On => MonitoringState.On,
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    public static FiberState ToProto(this Iit.Fibertest.Dto.FiberState state)
    {
        return state switch
        {
            Iit.Fibertest.Dto.FiberState.NotInTrace => FiberState.NotInTrace,
            Iit.Fibertest.Dto.FiberState.NotJoined => FiberState.NotJoined,
            Iit.Fibertest.Dto.FiberState.Unknown => FiberState.Unknown,
            Iit.Fibertest.Dto.FiberState.NotInZone => FiberState.NotInZone,
            Iit.Fibertest.Dto.FiberState.Ok => FiberState.Ok,
            Iit.Fibertest.Dto.FiberState.Suspicion => FiberState.Suspicion,
            Iit.Fibertest.Dto.FiberState.Minor => FiberState.Minor,
            Iit.Fibertest.Dto.FiberState.Major => FiberState.Major,
            Iit.Fibertest.Dto.FiberState.Critical => FiberState.Critical,
            Iit.Fibertest.Dto.FiberState.User => FiberState.User,
            Iit.Fibertest.Dto.FiberState.FiberBreak => FiberState.FiberBreak,
            Iit.Fibertest.Dto.FiberState.NoFiber => FiberState.NoFiber,
            Iit.Fibertest.Dto.FiberState.HighLighted => FiberState.HighLighted,
            Iit.Fibertest.Dto.FiberState.DistanceMeasurement => FiberState.DistanceMeasurement,
            Iit.Fibertest.Dto.FiberState.Nothing => FiberState.Nothing,
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    public static Iit.Fibertest.Dto.FiberState FromProto(this FiberState state)
    {
        int code = (int)state;
        return (Iit.Fibertest.Dto.FiberState)code;
    }

    public static TceLinkState ToProto(this TraceToTceLinkState state)
    {
        return state switch
        {
            TraceToTceLinkState.NoLink => TceLinkState.NoLink,
            TraceToTceLinkState.LinkTceOff => TceLinkState.SnmpTrapOff,
            TraceToTceLinkState.LinkTceOn => TceLinkState.SnmpTrapOn,
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    public static ChannelEvent ToProto(this Iit.Fibertest.Dto.ChannelEvent evnt)
    {
        return evnt switch
        {
            Iit.Fibertest.Dto.ChannelEvent.Nothing => ChannelEvent.Nothing,
            Iit.Fibertest.Dto.ChannelEvent.Broken => ChannelEvent.Broken,
            Iit.Fibertest.Dto.ChannelEvent.Repaired => ChannelEvent.Repaired,
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    public static BaseRefType ToProto(this Iit.Fibertest.Dto.BaseRefType type)
    {
        return type switch
        {
            Iit.Fibertest.Dto.BaseRefType.None => BaseRefType.None,
            Iit.Fibertest.Dto.BaseRefType.Precise => BaseRefType.Precise,
            Iit.Fibertest.Dto.BaseRefType.Fast => BaseRefType.Fast,
            Iit.Fibertest.Dto.BaseRefType.Additional => BaseRefType.Additional,
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    public static Iit.Fibertest.Dto.BaseRefType FromProto(this BaseRefType type)
    {
        int code = (int)type;
        return (Iit.Fibertest.Dto.BaseRefType)code;
    }

    public static EventStatus ToProto(this Iit.Fibertest.Dto.EventStatus status)
    {
        return status switch
        {
            Iit.Fibertest.Dto.EventStatus.JustMeasurementNotAnEvent => EventStatus.JustMeasurementNotAnEvent,
            Iit.Fibertest.Dto.EventStatus.EventButNotAnAccident => EventStatus.EventButNotAnAccident,
            Iit.Fibertest.Dto.EventStatus.NotImportant => EventStatus.NotImportant,
            Iit.Fibertest.Dto.EventStatus.Planned => EventStatus.Planned,
            Iit.Fibertest.Dto.EventStatus.NotConfirmed => EventStatus.NotConfirmed,
            Iit.Fibertest.Dto.EventStatus.Unprocessed => EventStatus.Unprocessed,
            Iit.Fibertest.Dto.EventStatus.Suspended => EventStatus.Suspended,
            Iit.Fibertest.Dto.EventStatus.Confirmed => EventStatus.Confirmed,
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    public static MonitoringCurrentStep ToProto(this Iit.Fibertest.Dto.MonitoringCurrentStep step)
    {
        return step switch
        {
            Iit.Fibertest.Dto.MonitoringCurrentStep.Unknown => MonitoringCurrentStep.Unknown,
            Iit.Fibertest.Dto.MonitoringCurrentStep.Idle => MonitoringCurrentStep.Idle,
            Iit.Fibertest.Dto.MonitoringCurrentStep.Toggle => MonitoringCurrentStep.Toggle,
            Iit.Fibertest.Dto.MonitoringCurrentStep.Measure => MonitoringCurrentStep.Measure,
            Iit.Fibertest.Dto.MonitoringCurrentStep.FailedOtauProblem => MonitoringCurrentStep.FailedOtauProblem,
            Iit.Fibertest.Dto.MonitoringCurrentStep.FailedOtdrProblem => MonitoringCurrentStep.FailedOtdrProblem,
            Iit.Fibertest.Dto.MonitoringCurrentStep.Interrupted => MonitoringCurrentStep.Interrupted,
            Iit.Fibertest.Dto.MonitoringCurrentStep.Analysis => MonitoringCurrentStep.Analysis,
            Iit.Fibertest.Dto.MonitoringCurrentStep.MeasurementFinished => MonitoringCurrentStep.MeasurementFinished,
            _ => throw new ArgumentOutOfRangeException()

        };
    }

    public static OpticalAccidentType ToProto(this Iit.Fibertest.Dto.OpticalAccidentType type)
    {
        return type switch
        {
            Iit.Fibertest.Dto.OpticalAccidentType.Break => OpticalAccidentType.Break,
            Iit.Fibertest.Dto.OpticalAccidentType.Loss => OpticalAccidentType.Loss,
            Iit.Fibertest.Dto.OpticalAccidentType.Reflectance => OpticalAccidentType.Reflectance,
            Iit.Fibertest.Dto.OpticalAccidentType.LossCoeff => OpticalAccidentType.LossCoeff,
            Iit.Fibertest.Dto.OpticalAccidentType.TotalLoss => OpticalAccidentType.TotalLoss,
            Iit.Fibertest.Dto.OpticalAccidentType.None => OpticalAccidentType.None,
            _ => throw new ArgumentOutOfRangeException()
        };
    }
}